Framework/Beamline/src/DetectorInfo.cpp

```diff
@@ -78,28 +90,251 @@ bool DetectorInfo::isEquivalent(const DetectorInfo &other) const {
+      const auto &interval2 = (*m_scanIntervals)[linearIndex2];
+      if (interval1 == interval2) {
+        if ((*m_isMasked)[linearIndex2] != (*other.m_isMasked)[linearIndex1])
+          failMerge("matching scan interval but mask flags differ");
+        if ((*m_positions)[linearIndex2] != (*other.m_positions)[linearIndex1])
+          failMerge("matching scan interval but positions differ");
+        if ((*m_rotations)[linearIndex2].coeffs() !=
+            (*other.m_rotations)[linearIndex1].coeffs())
+          failMerge("matching scan interval but rotations differ");
+        merge[linearIndex1] = false;
+      } else if ((interval1.first < interval2.second) &&
+                 (interval1.second > interval2.first)) {
+        failMerge("scan intervals overlap but not identical");
+      }
+    }
+  }
+
+  if (!m_scanCounts)
+    initScanCounts();
+  if (!m_indexMap)
+    initIndices();
+  // Temporary to accumulate scan counts (need original for index offset).
+  auto scanCounts(m_scanCounts);
+  for (size_t linearIndex = 0; linearIndex < other.m_positions->size();
+       ++linearIndex) {
+    if (!merge[linearIndex])
+      continue;
+    auto newIndex = getIndex(other.m_indices, linearIndex);
+    const size_t detIndex = newIndex.first;
+    newIndex.second += scanCount(detIndex);
+    scanCounts.access()[detIndex]++;
+    m_indexMap.access()[detIndex].push_back((*m_indices).size());
+    m_indices.access().push_back(newIndex);
+    m_isMasked.access().push_back((*other.m_isMasked)[linearIndex]);
+    m_positions.access().push_back((*other.m_positions)[linearIndex]);
+    m_rotations.access().push_back((*other.m_rotations)[linearIndex]);
+    m_scanIntervals.access().push_back((*other.m_scanIntervals)[linearIndex]);
+  }
+  m_scanCounts = std::move(scanCounts);
+}
+
+/// Returns the linear index for a pair of detector index and time index.
+size_t DetectorInfo::linearIndex(const std::pair<size_t, size_t> &index) const {
+  // The most common case are beamlines with static detectors. In that case the
+  // time index is always 0 and we avoid expensive map lookups. Linear indices
+  // are ordered such that the first block contains everything for time index 0
+  // so even in the time dependent case no translation is necessary.
+  if (index.second == 0)
+    return index.first;
+  return (*m_indexMap)[index.first][index.second];
+}
+
+/// Throws if this has time-dependent data.
+void DetectorInfo::checkNoTimeDependence() const {
+  if (isScanning())
+    throw std::runtime_error("DetectorInfo accessed without time index but the "
+                             "beamline has time-dependent (moving) detectors.");
+}
+
+void DetectorInfo::initScanCounts() {
+  checkNoTimeDependence();
+  m_scanCounts = Kernel::make_cow<std::vector<size_t>>(size(), 1);
+}
+
+void DetectorInfo::initScanIntervals() {
+  checkNoTimeDependence();
+  m_scanIntervals = Kernel::make_cow<std::vector<std::pair<int64_t, int64_t>>>(
```

Comment:
I suppose there are two alternatives:
- Store a vector of unique intervals and for each `(detector index, time index)` pair we store an index into that vector of unique intervals. This would save us `8 Byte` per spectrum.
- Store a vector of unique vectors of intervals. For each detector index we store an index into that vector. This saves use `8 * scanCount Byte` per spectrum.

The cost is some extra complexity. The memory saving might be negligible compared to the histogram or event list stored for each spectrum.

I do not mind doing this if it turns out that the current performance is affected significantly by having scan intervals for each individual detector, but I would prefer waiting until we know, i.e., did some benchmarking for merging workspaces that contain real amounts of data.